### PR TITLE
Tcp ip input get data and constructor

### PR DESCRIPTION
**Pull request by mehow93**
class TcpIpInput:
- constructor added,
- getData() added

main.cpp:
-one simply testcase added

### CODE REVIEW

File: TcpIpInput.h [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
 class TcpIpInput : public Input {
 private:
     short mPort;
+    struct sockaddr_in mAddress;
+    char mBuffer[1024] = {0};
+    int mServerFile;
 
 public:
-    TcpIpInput(short port);
+    TcpIpInput(short port, unsigned short domain, unsigned int type);
```
Comment (knopers8) [inline TcpIpInput.h]: Could you add a short explanation for `domain` and `type` arguments in a comment above the declaration? Also, I guess that in 99% the same values for `domain` and `type` would be used, so could you add some defaults?

Comment (knopers8) [reply TcpIpInput.h]: Another problem is that `port` is `short`, which limits the port number to 32767, but TCP ports can reach 65535. This is why `htons` takes uint16_t. Can you fix the data type?

File: TcpIpInput.h [hunk 2]
```diff
@@ -1,18 +1,28 @@
 #ifndef TCPIPINPUT_H
 #define TCPIPINPUT_H
 #include "Input.h"
+#include <unistd.h>
+#include <stdio.h>
+#include <sys/socket.h>
+#include <stdlib.h>
+#include <netinet/in.h>
+#include <string.h>
```
Comment (knopers8) [inline TcpIpInput.h]: Please include only the headers you need here. The rest can go to `TcpIpInput.cxx`, so they aren't brought to each file which includes `TcpIpInput.h`

File: TcpIpInput.cpp [hunk 3]
```diff
@@ -2,15 +2,59 @@
...
+    }
 }
 
 std::string TcpIpInput::getData(){
-    return "defualtValue";
+    int newSocket, valRead;
+    int addrLen = sizeof(this->mAddress);
+    if (listen(this->mServerFile, 3) < 0)
+    {
+        perror("listen");
```
Comment (knopers8) [inline TcpIpInput.cpp:43]: What is perror?

File: TcpIpInput.h [hunk 4]
```diff
@@ -1,18 +1,28 @@
...
 
 
 
 
 class TcpIpInput : public Input {
 private:
     short mPort;
+    struct sockaddr_in mAddress;
+    char mBuffer[1024] = {0};
```
Comment (knopers8) [inline TcpIpInput.h]: You should make the size a constant.

File: main.cpp [hunk 5]
```diff
@@ -258,12 +259,17 @@ TEST(fileInputClassTest, fullTest){
...
     std::unique_ptr<Input> t1 = std::make_unique<StdinInput>();
     EXPECT_EQ("T est\ne s t\ns t\n\n\n",t1->getData());
-}
+}*/
 
+TEST(TcpIpInputClassTest, virtualGetDataTest){
+    std::unique_ptr<Input> t1 = std::make_unique<TcpIpInput>(8080,AF_INET,INADDR_ANY);
+    t1->getData();
+    EXPECT_EQ(1,1);
+}
```
Comment (knopers8) [inline main.cpp:272]: Could you test if `getData()` can be invoked multiple times? I am not 100% sure if it would work.

**Review by knopers8 — CHANGES_REQUESTED** [review]
It's good overall, but a few issues should be fixed, see the comments.

File: TcpIpInput.h (hunk 1 above)
Comment (mehow93) [reply TcpIpInput.h]: done

File: TcpIpInput.cpp (hunk 3 above)
Comment (mehow93) [reply TcpIpInput.cpp:43]: it;s from stdio.h. It prints understandable code error. I'll add comment

Comment (knopers8) [reply TcpIpInput.cpp:43]: no need, it's fine. I just have never seen it before. Though, if you this is what you want, then perhaps you should use it in the whole project for printing errors, for consistency.

File: main.cpp (hunk 5 above)
Comment (mehow93) [reply main.cpp:272]: yup, unfortunatelly, it doesn't work in this case. I must do bugfix

File: TcpIpInput.cpp (hunk 3 above)
Comment (mehow93) [reply TcpIpInput.cpp:43]: OK

**Review by knopers8 — APPROVED** [review]

**Comment by knopers8** [comment]
We agreed that the remaining issue with invoking `getData()` several times will be addressed in another PR, merging.

**Pull request merged**